accounting/transferwise/tests/test_tasks.py

```diff
@@ -0,0 +1,92 @@
+#
+
+"""
+TransferWise tasks tests.
+"""
+
+from unittest import mock
+import datetime
+
+from django.core import mail
+from django.core.mail.message import EmailMessage
+from django.utils import timezone
+import ddt
+import freezegun
+
+from accounting.bank.tests.factories import BankAccountFactory
+from accounting.common.tests.base import TestCase
+from accounting.transferwise import models, tasks
+from accounting.transferwise.tests import factories
+
+NOW = datetime.datetime(2018, 1, 10, 20, 31, 3, 350993, tzinfo=timezone.utc)
+PAST = NOW - datetime.timedelta(days=NOW.day)
+
+
+@ddt.ddt
+class GenerateBulkPaymentCsvTestCase(TestCase):
+    """ Test cases for `tasks.generate_bulk_payment_csv`. """
+
+    @freezegun.freeze_time(NOW)
+    def setUp(self):
+        """ Set up test objects. """
+        self.create_client_and_provider_links()
+
+        # The providers need valid bank accounts  to get paid.
+        self.bank_account1 = BankAccountFactory(user_account=self.provider1)
+        self.bank_account2 = BankAccountFactory(user_account=self.provider2)
+
+        # Create some unpaid invoices that need to be paid in bulk.
+        self.invoice1 = factories.InvoiceFactory(provider=self.provider1, client=self.client, paid=False)
+        self.invoice2 = factories.InvoiceFactory(provider=self.provider2, client=self.client, paid=False)
+        super().setUp()
+
+    @mock.patch.object(models.TransferWiseBulkPayment, 'upload_to_google_drive')
+    @mock.patch.object(EmailMessage, 'attach_file')
+    @mock.patch.object(models.TransferWiseBulkPayment, 'to_bulk_payment_csv')
+    def test_generate_bulk_payment_csv(self, mock_to_bulk_payment_csv, mock_attach_file, mock_upload_to_google_drive):
+        """ A bulk payment CSV is generated for all unpaid invoices and sent off to the bulk payment sender. """
+        # Mock.
+        csv_path = 'https://drive.google.com/bulk_payment.csv'
+        mock_upload_to_google_drive.return_value = {'webContentLink': csv_path}
+        mock_to_bulk_payment_csv.return_value = mock.MagicMock(name=csv_path)
+
+        # Call.
+        tasks.generate_bulk_payment_csv()
+
+        # The invoices will be marked as paid.
+        self.invoice1.refresh_from_db()
+        self.invoice2.refresh_from_db()
+        self.assertTrue(self.invoice1.paid)
+        self.assertTrue(self.invoice2.paid)
+
+        # 1 sender, so 1 call to everything.
+        mock_to_bulk_payment_csv.assert_called_once()
+        mock_attach_file.assert_called_once()
+        mock_upload_to_google_drive.assert_called_once()
+        self.assertEqual(len(mail.outbox), 1)
+
+        # 1 bulk payment that contains 2 payments, 1 for each provider.
+        self.assertEqual(models.TransferWiseBulkPayment.objects.all().count(), 1)
+        self.assertEqual(models.TransferWisePayment.objects.all().count(), 2)
+        self.assertEqual(models.TransferWiseBulkPayment.objects.first().payments.all().count(), 2)
```

Comment:
@UmanShahzad I think you can simplify the code above by calling `count()` on the manager (instead of calling `all()` first). Also, this test expects there to be exactly one `TransferWiseBulkPayment`, so we could use `get` instead of `first`:

```python
        self.assertEqual(models.TransferWiseBulkPayment.objects.count(), 1)
        self.assertEqual(models.TransferWisePayment.objects.count(), 2)
        self.assertEqual(models.TransferWiseBulkPayment.objects.get().payments.count(), 2)
```

Reply:
Oh yes, good catches :+1: